File: plugin/src/shared/sc/plugin2019/Direction.java
package sc.plugin2019;

import sc.plugin2019.util.Point;

/** Bewegungsrichtung für Move */
public enum Direction {
  UP,
  UP_RIGHT,
  RIGHT,
  DOWN_RIGHT,
  DOWN,
  DOWN_LEFT,
  LEFT,
  UP_LEFT;

  public Point shift() {
    int shiftX = 0;
    int shiftY = 0;
    switch (this) {
      case UP_RIGHT:
        shiftX = 1;
      case UP:
        shiftY = 1;
        break;
      case DOWN_RIGHT:
        shiftY = -1;
      case RIGHT:
        shiftX = 1;
        break;
      case DOWN_LEFT:
        shiftX = -1;
      case DOWN:
        shiftY = -1;
        break;
      case UP_LEFT:
        shiftY = 1;
      case LEFT:
        shiftX = -1;
        break;
    }
    return new Point(shiftX, shiftY);
  }

}

File: sdk/src/main/server-api/sc/api/plugins/Direction.kt
package sc.api.plugins

import com.thoughtworks.xstream.annotations.XStreamAlias
import kotlin.random.Random

@XStreamAlias(value = "direction")
enum class Direction(val vector: Vector): IVector by vector {
    UP(Vector(0, 1)),
    UP_RIGHT(Vector(1, 1)),
    RIGHT(Vector(1, 0)),
    DOWN_RIGHT(Vector(1, -1)),
    DOWN(Vector(0, -1)),
    DOWN_LEFT(Vector(-1, -1)),
    LEFT(Vector(-1, 0)),
    UP_LEFT(Vector(1, -1));
    
    companion object {
        val diagonals = arrayOf(UP_RIGHT, DOWN_RIGHT, DOWN_LEFT, UP_LEFT)
        val cardinals = arrayOf(UP, RIGHT, DOWN, LEFT)
    }
    
    val opposite: IVector
        get() = Vector(-vector.dx, -vector.dy)
}

enum class HexDirection(val vector: Vector): IVector by vector {
    RIGHT(Vector(+2, 0)),
    DOWN_RIGHT(Vector(+1, +1)),
    DOWN_LEFT(Vector(-1, +1)),
    LEFT(Vector(-2, 0)),
    UP_LEFT(Vector(-1, -1)),
    UP_RIGHT(Vector(+1, -1));
    
    fun withNeighbors(): Array<HexDirection> = arrayOf(rotatedBy(-1), this, rotatedBy(1))
    
    fun opposite(): HexDirection = values().let { it[(ordinal + 3) % it.size] }
    
    fun turnCountTo(target: HexDirection): Int {
        val diff = target.ordinal - this.ordinal
        return if(diff >= 0) diff else diff + values().size
    }
    
    fun rotatedBy(turns: Int): HexDirection = values().let { it[(ordinal + turns).mod(it.size)] }
    
    companion object {
        fun random(): HexDirection = values()[Random.nextInt(values().size)]
    }
}
